### ISSUE TITLE
Error 1007 on restricted network with ipv6 (not only) with TURN

### ISSUE THREAD

**Issue by unicode-it** [bigbluebutton/bigbluebutton#9514]
**Describe the bug**

**To Reproduce**

1. Restrict all outgoing Traffic except ports 53 80 443
2. Make sure to have ipv6 connectivity
3. Try to join conference with audio
4. See error 1007

**It is reproducible on official BBB demo!**

**Expected behavior**
Connect to conference with audio/micro

**Actual behavior**
No Microphone connection possible (1007)
Also no audio without microphone but no error 

**Desktop (please complete the following information):**
 - OS: Linux (Arch / Ubuntu)
 - Browser: Chrome, Firefox 

**Additional context**
Browser log attached
[BBB_Browserlog.txt](https://github.com/bigbluebutton/bigbluebutton/files/4617796/BBB_Browserlog.txt)

**Comment by nutton** [comment]
>ERROR: clientLogger: ICE connection failed before success
did you have installed/configured stun/turn server for freeswitch/kurento/bbb-web?
did you opened/forwarded udp range in firewall for kurento/freeswitch?
https://docs.bigbluebutton.org/2.2/setup-turn-server.html
https://docs.bigbluebutton.org/2.2/configure-firewall.html

**Comment by nutton** [comment]
also you may be needed incoming udp traffic in range ~ 20000-65535 ports, because audio/video goes thru udp transport.

**Comment by unicode-it** [comment]
Yes, we use our own stun/turn server on an other machine. 
Firewall is open to port Range 16384-32768 UDP.

**Comment by nutton** [comment]
Also you can verify which candidates are anounced through your stun/turn server.
https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/
and turn/stun ports should be opened/forwarded(if nat used).
make sure that you used right credentials, certs ports and URI between bbb-stack and your stun/turn server.

**Comment by unicode-it** [comment]
Thank you for your advice, but as we wrote, it is also not working on the official BBB Demo https://demo.bigbluebutton.org/gl/ 

we are pretty sure the turn server works, as it is working w/o problems with ipv4 only and restricted networks

**Comment by ralferlebach** [comment]
I had the same issue and it turned out that it is a matter on the user's side (restiction in the router, UDP ports are blocked) So it's not a bug or problem of your server configuration or BBB at all.

As a working solution, you may use VPN tunneling and advice your users behind such restricted network settings to make use of that option too.

Greetings

**Comment by unicode-it** [comment]
Sorry, i dont see this as an issue on the user side. For our understanding one functionality of TURN/Stun is, to make it possible to take part on the audio conference for people who are connecting from restricted networks. 

We also investigated with many people who are encountering the problem from sucht type of networks, like guest wlans or company networks. 

We simulated the behavior with restricting our client firewall, to collect some knowledge about the issue. As soon as you have an ipv6, or ipv6/ipv4 connection and only the ports 80/443 and 53, you cannot connect anymore.

**Comment by ichdasich** [comment]
Can you elaborate a little more on the restrictions? When I was debugging a user recently, I created a similar setting (53/80/443 only on a dual-stack network), and things were just working fine.

Specifically:
- Does this behavior change when you remove the v6 address from the interface (you explicitly mention it in the title)
- Do you see (when using firefox) any candidates on the clients?
- is 53/80/443 outbound fully routed, or is there also a HTTP proxy in the mix?

**Comment by nutton** [comment]
you sure that you see all candidates(ip4/6),even directly declared via bbb-web?
check with netcat/tcpdump on both sides (server,client) tcp/udp traffic on all declared ports, endpoints correctly routed/filtered?
although coturn, as recomended  by official manual, supports ipv6.
in turn-stun-servers.xml you can forcely point to your endpoints, both 4|6.
maybe, i'm not sure - not much experience with paralel stacked setups, but maybe you need to do dual setup for all of bbb-stack(mostly for coturn, freeswitch, kurento and may be some others).

**Comment by ichdasich** [comment]
Hm, and: I did not encounter this when debugging my turn (simulated the exact same setting) on my infrastructure. Are your TURN/BBB systems dualstack as well? (demo. is certainly only legacy IP).

**Comment by maaeps** [comment]
I have the same issue in a restricted network. For testing purposes I have now disabled UDP on my working PC (in a "normal" network).

On Firefox, when joining a meeting (on https://test.bigbluebutton.org or https://demo1.bigbluebutton.org or my down server) and then clicking once on "Listen only", `ICE failed, your TURN server appears to be broken, see about:webrtc for more details` is printed in the console. After a few more seconds, I am asked, if I want allow the site to use the microphone. When agreeing, everything behaves as normal.

On Safari this problem does not occur, but I can't share my webcam (Error 1020) and I can not see the shared screen / the shared webcams (spinning forever).

**Comment by unicode-it** [comment]
> Hm, and: I did not encounter this when debugging my turn (simulated the exact same setting) on my infrastructure. Are your TURN/BBB systems dualstack as well? (demo. is certainly only legacy IP).

Yes everything is dual-stack. Concerning the bbb server and the complexity of components I can only say it is Dual-Stack according to the official documentation.

I found some time to debug this further as one of our customers now hits the issue. With the client configuration as mentioned above (dual ipv4, ipv6 and 53/80/443 only) there is no traffic (not a single package) to the turn server on ipv6 (tested with tcpdump on the client). With ipv4 only a lot of traffic is captured. 
I checked for the turn server:
- DNS A and AAAA are correctly set
- port 443 on ipv6 addr is open and responding

It seems on ipv6 either the necessity for turn relay is not recognized or wrong configuration is passed to the browser on connection negotiation. I could provide data from the browser-console if anybody has an idea.

**Comment by ichdasich** [comment]
Yes, the webrtc stats from FF would be nice. (about:webrtc)

**Comment by jodoma** [comment]
I may or may not face the same/a similar issue: 
https://groups.google.com/forum/?utm_source=footer#!msg/bigbluebutton-setup/OzjwjFe7Ksk/wYOMP3_sBQAJ

Let me know if you still need webrtc stats or any logs.

**Comment by deg0nz** [comment]
I had the exact same issue. Tests with Firefox and `media.peerconnection.ice.relay_only=true` always failed with error 1007 and some clients had trouble connecting.

BBB and coturn have v4 and v6 enabled as well as the client.

After a lot of research, I came across this configuration in the BBB docs: https://docs.bigbluebutton.org/install#configure-bigbluebuttonfreeswitch-to-support-ipv6

After setting the external v4 and v6 addresses of my server in the Nginx config, Firefox was able to connect without any issues. (This was my only method to check the connection via relay though...)

Can anyone here confirm this? Or tell, if they still have the problems even _with_ the Nginx config?

**Comment by ffdixon** [comment]
> Can anyone here confirm this?

Confirmed.  This is the line to add in nginx to support incoming IPV6 websocket connections for FreeSWITCH.

**Comment by unicode-it** [comment]
@deg0nz Thx a lot for pointing to the section of the docs. I cannot test the issue by myself at the moment, because I am on a site without ipv6, but I rolled it out to a customer complaining about the issue for a while and it seems to have solved it.

As we implemented quit some changes to support ipv6 in 2020 and it also didn't work on the official demo when we opened the issue, I wonder if these instructions were added to the docs later.

I also wonder how it solves the issue as the modified `proxy_pass` directive shouldn't have changed anything for the client but is just a reverse proxy on the server to reach freeswitch on the same host. But maybe the missing `enable-3pcc` was the change to fix that behavior. 

After all I am very happy about the solution and the support.

**Comment by prlanzarin** [comment]
> I also wonder how it solves the issue as the modified proxy_pass directive shouldn't have changed anything for the client but is just a reverse proxy on the server to reach freeswitch on the same host. 

FreeSWITCH dual stack handling is weird. It has two separate external profiles for IPv4 and IPv6. There's no unified profile. 
The way it infers the desired profile is via the signaling transport IP version. So if signaling transport is bound to IPv4 -> profile is IPv4 and media/UDP connectivity via IPv6 will not be possible (and vice-versa).
Hence why the reverse proxy has to propagate the IP version.

> But maybe the missing enable-3pcc was the change to fix that behavior.

Shouldn't have any effect on this case AFAIK. SIP 3PCC was (is) enabled in BBB's deploy of FS due to a quirk FS <= 1.8 had where INVITE requests without SDPs wouldn't work. Enabling 3PCC generated a (beneficial) side effect that made the call flow I wanted (which is valid) work. Whatever worked back then. AFAIK >=1.10 might have this fixed so maybe the flag isn't necessary anymore.

**Comment by Daniel-at-git** [comment]
I have a similar problem. With Firefox and activated "media.peerconnection.ice.relay_only" it's working fine. If I just allow 53,80,443 it's not working with Chrome. Tested with my own (2.5.4) system and test.bigbluebutton.org

If I remove
keep-address-family
from turn.conf it's working fine with Chrome.

I posted on bigbluebutton-dev but didn't get an answer
https://groups.google.com/g/bigbluebutton-dev/c/n2Vb2zme1uY
